packages/Caching/ValueObject/Storage/FileCacheStorage.php

```diff
@@ -24,59 +23,60 @@ public function __construct(
-
-        $filePath = $cacheFilePaths->getFilePath();
-        if (! is_file($filePath)) {
-            return null;
-        }
-
-        $cacheItem = require $filePath;
-        if (! $cacheItem instanceof CacheItem) {
-            return null;
-        }
-
-        if (! $cacheItem->isVariableKeyValid($variableKey)) {
-            return null;
-        }
-
-        return $cacheItem->getData();
+        return (function (string $key, string $variableKey) {
+            $cacheFilePaths = $this->getCacheFilePaths($key);
+
+            $filePath = $cacheFilePaths->getFilePath();
+            if (!\is_file($filePath)) {
+                return null;
+            }
+            $cacheItem = (require $filePath);
+            if (!$cacheItem instanceof CacheItem) {
+                return null;
+            }
+            if (!$cacheItem->isVariableKeyValid($variableKey)) {
+                return null;
+            }
+            return $cacheItem->getData();
+        })($key, $variableKey);
     }
 
     /**
+     * @param string $key
+     * @param string $variableKey
      * @param mixed $data
+     * @return void
      */
-    public function save(string $key, string $variableKey, $data): void
+    public function save(string $key, string $variableKey, $data) : void
     {
         $cacheFilePaths = $this->getCacheFilePaths($key);
-
         $this->smartFileSystem->mkdir($cacheFilePaths->getFirstDirectory());
         $this->smartFileSystem->mkdir($cacheFilePaths->getSecondDirectory());
+        $path = $cacheFilePaths->getFilePath();
 
-        $tmpPath = sprintf('%s/%s.tmp', $this->directory, Random::generate());
-        $errorBefore = error_get_last();
-        $exported = @var_export(new CacheItem($variableKey, $data), true);
-        $errorAfter = error_get_last();
-
+        $tmpPath = \sprintf('%s/%s.tmp', $this->directory, Random::generate());
+        $errorBefore = \error_get_last();
+        $exported = @\var_export(new CacheItem($variableKey, $data), true);
+        $errorAfter = \error_get_last();
         if ($errorAfter !== null && $errorBefore !== $errorAfter) {
-            $errorMessage = sprintf(
-                'Error occurred while saving item "%s" ("%s") to cache: "%s"',
-                $key,
-                $variableKey,
-                $errorAfter['message']
-            );
-            throw new CachingException($errorMessage);
+            throw new CachingException(\sprintf('Error occurred while saving item %s (%s) to cache: %s', $key, $variableKey, $errorAfter['message']));
+        }
+        // for performance reasons we don't use SmartFileSystem
+        FileWriter::write($tmpPath, \sprintf("<?php declare(strict_types = 1);\n\nreturn %s;", $exported));
+        $renameSuccess = @\rename($tmpPath, $path);
+        if ($renameSuccess) {
+            return;
+        }
+        @\unlink($tmpPath);
+        if (\DIRECTORY_SEPARATOR === '/' || !\file_exists($path)) {
```

Comment:
have you tried in different OS ? `DIRECTORY_SEPARATOR` is `/` in `macOS`:

```bash
➜  ~ php -a
Interactive shell

php > echo DIRECTORY_SEPARATOR;
/

```

Reply:
this implementation is taken 1:1 from phpstan.

I guess this condition is meant as a non-windows feature detect

Reply:
That possibly a dead code then, `true || false` will result true, `true || true` will result true, see https://3v4l.org/pKcY1

Reply:
I can't tell the details why phpstan has implemented it like that.

Reply:
Some comment would be useful so we don't use this blindly in the future